### PR TITLE
Add rate providers example

### PR DESCRIPTION
**Pull request by lucekdudek**
As I want to run rating on every provider that made a proposal I need a way to stop computing after getting proposals in `demand.initial_proposals(),`. The most suited interface for it is the `set_no_more_children` method. It lead me to that construction:
```python
async def delay(delay: timedelta, func: Callable):
    await asyncio.sleep(delay.total_seconds())
    return func()
...
asyncio.create_task(delay(timedelta(seconds=5), demand.set_no_more_children))
```

---

I run into problem in which my tasks were failing for unknown reason and I had trouble to get logs that would explain what is the problem. I ended up with this error handling:
```python
try:
        batch = await activity.execute_commands(
            commands.Run(["/golem/entrypoints/run-blender.sh"]),
        )
        await batch.wait(timeout=20)
    except BatchError as err:
        raise BatchError(
            batch=err.batch,
            msg=f"Rating failed on {err}\nCommand stdout: {batch.events[-1].stdout}\nCommand stderr: {batch.events[-1].stderr}",
        )
```

This produces output like this:
```bash
Activity Activity(1d007bcb48df4b7199fc358aede11b1b) failed because of:
Rating failed on PoolingBatch(9ff9229a147746389d0cbf1a63f2b17e) failed on command 1: ExeScript command exited with code 1
Command stdout: None
Command stderr: Traceback (most recent call last):
  File "/golem/entrypoints/render_entrypoint.py", line 19, in <module>
    run_blender_task()
  File "/golem/entrypoints/render_entrypoint.py", line 15, in run_blender_task
    results_info = blender.render(params, paths)
  File "/golem/entrypoints/scripts/render_tools/blender_render.py", line 158, in render
    crop_counter)
  File "/golem/entrypoints/scripts/render_tools/blender_render.py", line 101, in gen_blender_script_file
    outfilebasename = crop["outfilebasename"]
KeyError: 'outfilebasename'
```

---

I ended up adding additional example showcasing example handling in current version of golem-core-python

### CODE REVIEW

File: examples/exception_handling/exception_handling.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    loop = asyncio.get_event_loop()
+    task = loop.create_task(main())
+    try:
+        loop.run_until_complete(task)
+    except KeyboardInterrupt:
+        task.cancel()
+        try:
+            loop.run_until_complete(task)
+        except asyncio.CancelledError:
+            pass
```
Comment (approxit) [inline examples/exception_handling/exception_handling.py:90]: This whole block is handled by default in [`asyncio.run`](https://github.com/python/cpython/blob/3.10/Lib/asyncio/runners.py#L47), even with cancellation attempt.

```suggestion
    asyncio.run(main())
```

This pattern should be even propagated everywhere in our code.

File: examples/rate_providers/rate_providers.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    )
+    await batch.wait(timeout=60)
+    return activity
+
+
+async def rate_provider(activity: Activity) -> str:
+    provider_id = activity.parent.parent.data.issuer_id
+    print(f"Rating {provider_id}")
+    frame_config = FRAME_CONFIG_TEMPLATE.copy()
+    fname = f"out{0:04d}.png"  # format used by `run-blender.sh` included in `PAYLOAD`
```
Comment (approxit) [inline examples/rate_providers/rate_providers.py:59]:
```suggestion
    fname = "out{0:04d}.png"  # format used by `run-blender.sh` included in `PAYLOAD`
```

As we're not want to call fstring right now, but later. I'm surprised that it's not raising exception due to lack of `0` variable in the scope.

File: examples/rate_providers/rate_providers.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    loop = asyncio.get_event_loop()
+    task = loop.create_task(main())
+    try:
+        loop.run_until_complete(task)
+    except KeyboardInterrupt:
+        task.cancel()
+        try:
+            loop.run_until_complete(task)
+        except asyncio.CancelledError:
+            pass
```
Comment (approxit) [inline examples/rate_providers/rate_providers.py:128]:
```suggestion
    asyncio.run(main())
```

Same as previous similar block.

File: examples/score_based_providers/score_based_providers.py [hunk 4]
```diff
@@ -131,4 +132,13 @@ async def main() -> None:
...
+    loop = asyncio.get_event_loop()
+    task = loop.create_task(main())
+    try:
+        loop.run_until_complete(task)
+    except KeyboardInterrupt:
+        task.cancel()
+        try:
+            loop.run_until_complete(task)
+        except asyncio.CancelledError:
+            pass
```
Comment (approxit) [inline examples/score_based_providers/score_based_providers.py:144]:
```suggestion
    asyncio.run(main())
```

Same as previous similar block.

**Review by approxit — CHANGES_REQUESTED** [review]
One minor thing, and few instances of the same case, where `asyncio.run()` will handle significantly better loop handling, with clean task cancellation and shutting down asyncgens.

File: examples/exception_handling/exception_handling.py (hunk 1 above)
Comment (lucekdudek) [reply examples/exception_handling/exception_handling.py:90]: ctr-c didn't work for me when I used `asyncio.run(main())`. I had to send it twice for some reason.

File: examples/rate_providers/rate_providers.py (hunk 2 above)
Comment (lucekdudek) [reply examples/rate_providers/rate_providers.py:59]: 0 is a number, can't make a variable starting with a number :joy:

File: examples/exception_handling/exception_handling.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+async def trigger_exception(activity: Activity) -> str:
+    provider_id = activity.parent.parent.data.issuer_id
+    try:
+        batch = await activity.execute_commands(
+            commands.Deploy(),
+            commands.Start(),
+            commands.Run(["/golem/entrypoints/run-blender.sh"]),
+        )
+    except BatchError:
```
Comment (johny-b) [inline examples/exception_handling/exception_handling.py:33]: I don't think you can get a `BatchError` here, it is raised in `batch.wait` not in `activity.execute_commands`

Comment (lucekdudek) [reply examples/exception_handling/exception_handling.py:33]: True, I tested it having `wait` inside try block

Comment (lucekdudek) [reply examples/exception_handling/exception_handling.py:34]: fixed

Comment (shadeofblue) [reply examples/exception_handling/exception_handling.py:34]: hm... still, what's the point of catching the exception only to raise it immediately without doing anything else?

File: examples/exception_handling/exception_handling.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+    Map,
+    default_create_activity,
+    default_create_agreement,
+    default_negotiate,
+)
+
+PAYLOAD = Payload.from_image_hash(
+    "9a3b5d67b0b27746283cb5f287c13eab1beaa12d92a9f536b747c7ae"
+)
+SUBNET = "public"
```
Comment (johny-b) [inline examples/exception_handling/exception_handling.py:22]: I think a better approach could be to just bump the required `yapapi` version. I tested this briefly and nothing unexpected happened.
I don't know about any reason for using `0.9.3` now.

File: examples/exception_handling/exception_handling.py [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+
+
+async def trigger_exception(activity: Activity) -> str:
```
Comment (johny-b) [inline examples/exception_handling/exception_handling.py:25]: I'm not sure what is the exact purpose of the example.
BUT if you want a "readable way of creating an exception in RUN command" i'd try something like

```
async def run_invalid_command(activity: Activity) -> None:
    batch = await activity.execute_commands(
        commands.Deploy(),
        commands.Start(),
        commands.Run("/ooops/no_such_command"),
    )
    await batch.wait(timeout=20)
````

But I might be missing the point of the example.

File: examples/rate_providers/rate_providers.py [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+
+    async with golem:
+        allocation = await golem.create_allocation(1.0)
+        payment_manager = DefaultPaymentManager(golem, allocation)
+        demand = await golem.create_demand(
+            PAYLOAD, allocations=[allocation], subnet=SUBNET
+        )
+
+        # `set_no_more_children` has to be called so `initial_proposals` will eventually stop yielding
+        asyncio.create_task(delay(timedelta(seconds=5), demand.set_no_more_children))
```
Comment (johny-b) [inline examples/rate_providers/rate_providers.py:98]: There are different solutions available to achieve something similar:
1. Just `break` out of the `async for result in chain` loop
2. Add a `Limit` somewhere
3. Wrap `demand.initial_proposals()` in something like
```
async def proposals(demand, end time):
    async for proposal in demand.initial_proposals():
        if datetime.now() >= end_time:
            return
        yield proposal
```

I'm curious why did you choose this particular solution?
I'm not saying it is wrong :) But I'm also not super-sure if `set_no_more_children` should be a default solution for such problems.

Comment (johny-b) [reply examples/rate_providers/rate_providers.py:98]: And regarding the comment: `demand.initial_proposals()` yields only when a value is requested.
Once you have processed 10 providers nothing more will be requested, so nothing more will be yielded.

**Review by approxit — APPROVED** [review]
No further comments from my side.

File: examples/exception_handling/exception_handling.py (hunk 6 above)
Comment (lucekdudek) [reply examples/exception_handling/exception_handling.py:22]: I don't think we want to bump yapapi in the middle of removing yappai dependency.

We have this task to handle issue of having this `SUBNET = "public"` in examples
https://github.com/golemfactory/golem-core-python/issues/73

File: examples/exception_handling/exception_handling.py (hunk 7 above)
Comment (lucekdudek) [reply examples/exception_handling/exception_handling.py]: Good point, added

**Pull request merged**